halo2_proofs/src/plonk/permutation/keygen.rs

```diff
@@ -108,60 +115,173 @@ impl Assembly {
+            return Err(Error::BoundsFailure);
         }
 
-        // Computes the permutation polynomial based on the permutation
-        // description in the assembly.
-        let mut permutations = vec![domain.empty_lagrange(); p.columns.len()];
-        {
-            parallelize(&mut permutations, |o, start| {
-                for (x, permutation_poly) in o.iter_mut().enumerate() {
-                    let i = start + x;
-                    for (j, p) in permutation_poly.iter_mut().enumerate() {
-                        let (permuted_i, permuted_j) = self.mapping[i][j];
-                        *p = deltaomega[permuted_i][permuted_j];
-                    }
-                }
-            });
+        // See book/src/design/permutation.md for a description of this algorithm.
+        let mut left_cycle = self.aux.get(&(left_column, left_row));
+        let mut right_cycle = self.aux.get(&(right_column, right_row));
+
+        // If left and right are in the same cycle, do nothing.
+        if (left_cycle == right_cycle) && (left_cycle.is_some()) && (right_cycle.is_some()) {
+            return Ok(());
+        }
+
+        let left_cycle_size = match left_cycle {
+            // this should unwrap safely
+            Some(i) => self.cycles[*i].len(),
+            None => 1,
+        };
+
+        let right_cycle_size = match right_cycle {
+            // this should unwrap safely
+            Some(i) => self.cycles[*i].len(),
+            None => 1,
+        };
+
+        if left_cycle_size < right_cycle_size {
+            std::mem::swap(&mut left_cycle, &mut right_cycle);
+            std::mem::swap(&mut left_column, &mut right_column);
+            std::mem::swap(&mut left_row, &mut right_row);
         }
 
-        // Pre-compute commitments for the URS.
-        let mut commitments = Vec::with_capacity(p.columns.len());
-        for permutation in &permutations {
-            // Compute commitment to permutation polynomial
-            commitments.push(
-                params
-                    .commit_lagrange(permutation, Blind::default())
-                    .to_affine(),
-            );
+        // extract cycle elements
+        let right_cycle_elems = match right_cycle {
+            Some(i) => {
+                let entry = self.cycles[*i].clone();
+                self.cycles[*i] = BTreeSet::new();
+                entry
+            }
+            None => [(right_column, right_row)].into(),
+        };
+
+        assert!(right_cycle_elems.contains(&(right_column, right_row)));
+
+        // merge cycles
+        let cycle_idx = match left_cycle {
+            Some(i) => {
+                let entry = &mut self.cycles[*i];
+                entry.extend(right_cycle_elems.clone());
+                *i
+            }
+            // if they were singletons -- create a new cycle entry
+            None => {
+                let mut set: BTreeSet<(usize, usize)> = right_cycle_elems.clone();
```

Comment:
My intuition is that this is slower than just not keeping track of the sizes (and so never swapping the left and right sets, or doing so based on a cheap heuristic such as whether each set has size > 1). Although that has worst-case complexity $O(n^2)$, that would only be an issue in practice if you have lots of cells in the same equality set.

Reply:
I've removed this -- ty for the tip :)